android/src/main/java/com/swmansion/reanimated/reflection/ReanimatedMapBuilder.java

```diff
@@ -147,19 +157,67 @@ public A asArray() {
     }
 
     @NonNull
-    public M asMap() {
-        return (M) mapContext.copy();
+    public M asMap(boolean useNativeBuilder) {
+        return ((M) (useNativeBuilder ? ReanimatedWritableNativeMap.fromMap(mapContext) : mapContext));
     }
 
     @NonNull
     public Object export() {
-        return isArray() ? asArray() : asMap();
+        return export(true);
+    }
+
+    @NonNull
+    public Object export(boolean useNativeBuilder) {
+        return isArray() ? asArray(useNativeBuilder) : asMap(useNativeBuilder);
     }
 
     @NonNull
     @Override
     public String toString() {
-        return export().toString();
+        return export(false).toString();
+    }
+
+    public static Object exportValue(Object value) {
+        if (value instanceof ReadableMap) {
+            return ReanimatedWritableNativeMap.fromMap((ReadableMap) value);
+        } else if (value instanceof ReadableArray) {
+            return ReanimatedWritableNativeArray.fromArray((ReadableArray) value);
+        } else {
+            return value;
+        }
+    }
+
+    public static class ReanimatedMapBuilderManager {
+
+        private final ReanimatedMapBuilder mMapBuilder;
+
+        public ReanimatedMapBuilderManager(ReanimatedMapBuilder mapBuilder) {
+            mMapBuilder = mapBuilder;
+        }
+
+        // TODO: 08/12/2019  setting value at path doesn't propagate to collection
+        public void set(ArrayList<Object> path, Object value) {
+            ReanimatedBridge.ReadableCollection collection = mMapBuilder;
+            Object key;
+            for (int i = 0; collection != null && i < path.size() - 1; i++) {
+                key = path.get(i);
+                collection = collection.has(key) ? collection.value(key, ReanimatedBridge.ReadableCollection.class) : null;
+            }
+
+            if (collection != null) {
+                key = path.get(path.size() - 1);
+                Object source = mMapBuilder.resolver(key).source();
+                if (source instanceof ReanimatedWritableMap) {
+                    ((ReanimatedWritableMap) source).put((String) key, value);
+                } else if (source instanceof ReanimatedWritableArray) {
+                    ((ReanimatedWritableArray) source).set((int) key, value);
```

Comment:
@osdnk I need your help here.
This method is supposed to help reduce map value calculation overload.
It manages to set the value on `source` at the right key. It seems that `source` is a local copy, not a ref to the original object, hence changes do not propagate, making this approach stale.